### PR TITLE
feat: self discover agent

### PR DESCRIPTION
**Pull request by sb-git-cloud**
**What is the problem that this fixes or functionality that this introduces? Does it fix any open issues?**

Implements a Self Discover agent, which provides a step-by-step plan to CodeActAgent. This is related to the open PR https://github.com/OpenDevin/OpenDevin/issues/46.

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**
The agent first chooses reasoning modules that may be relevant for the user-given task. Then it adapts those to the specific task and finally it creates a plan, which is then sent to CodeAct for execution. 

As of now it is in beta version.

---
**Other references**

### CODE REVIEW

**Comment by tobitege** [comment]
Thanks for building and contributing this! 🤗

File: agenthub/self_discover_agent/reasoning_action_parser.py
Comment (enyst) [inline agenthub/self_discover_agent/reasoning_action_parser.py:6]:
```suggestion
from .reasoning_action import REASONING_MODULE_LIST, NestedDict
```

File: agenthub/self_discover_agent/prompt.py
Comment (enyst) [inline agenthub/self_discover_agent/prompt.py:194]:
```suggestion
            reasoning_module_list=REASONING_MODULE_LIST,
```

File: agenthub/self_discover_agent/prompt.py
Comment (enyst) [inline agenthub/self_discover_agent/prompt.py:36]:
```suggestion
{reasoning_module_list}
```

File: agenthub/self_discover_agent/prompt.py
Comment (tobitege) [reply agenthub/self_discover_agent/prompt.py:194]: Really love this funny typo! 😂

File: agenthub/self_discover_agent/reasoning_action.py
Comment (enyst) [inline agenthub/self_discover_agent/reasoning_action.py:7]:
```suggestion
REASONING_MODULE_LIST: List[str] = [
```
Just some nitpick :)

File: agenthub/self_discover_agent/reasoning_action_parser.py
Comment (sb-git-cloud) [reply agenthub/self_discover_agent/reasoning_action_parser.py]: Thanks for the review. Typo is fixed now.

File: agenthub/self_discover_agent/prompt.py
Comment (sb-git-cloud) [reply agenthub/self_discover_agent/prompt.py]: Haha.
Fixed.

File: agenthub/self_discover_agent/prompt.py
Comment (sb-git-cloud) [reply agenthub/self_discover_agent/prompt.py]: Done.

File: agenthub/self_discover_agent/reasoning_action.py
Comment (sb-git-cloud) [reply agenthub/self_discover_agent/reasoning_action.py]: Fixed :)

File: agenthub/self_discover_agent/prompt.py
Comment (tobitege) [inline agenthub/self_discover_agent/prompt.py:22]: Just commenting: I wonder whether this instruction was _ever_ followed. 😂

File: agenthub/self_discover_agent/prompt.py
Comment (tobitege) [inline agenthub/self_discover_agent/prompt.py:77]: Do we want to keep this commented out code?

**Review by tobitege — COMMENTED** [review]
Would love to see some test for this, if possible. Maybe in a follow up PR.

File: agenthub/self_discover_agent/prompt.py
Comment (sb-git-cloud) [reply agenthub/self_discover_agent/prompt.py:22]: Yeah good question. I don't have any data to answer...

**Comment by li-boxuan** [comment]
Can you showcase some simple examples? Screenshots would suffice. Then maybe we could run some evaluation against it. C.C. @xingyaoww

File: agenthub/codeact_agent/codeact_agent.py
Comment (xingyaoww) [inline agenthub/codeact_agent/codeact_agent.py:198]: Is this absolutely necessary for CodeAct to be delegated by other agents? cc @li-boxuan

File: agenthub/self_discover_agent/agent.py
Comment (xingyaoww) [inline agenthub/self_discover_agent/agent.py:39]: It will be nice if we can use the newly implemented Jinja template for this agent - but no pressure - we can do it follow up PR.

https://github.com/OpenDevin/OpenDevin/blob/14a4e45cbbbb01eba01b47e5d874c2583decdfd8/agenthub/codeact_agent/codeact_agent.py#L76-L80

If the template is too rigid, we can figure out a way to generalize it a bit

File: agenthub/self_discover_agent/prompt.py
Comment (xingyaoww) [inline agenthub/self_discover_agent/prompt.py:189]: On a second thought: the prompt here is probably a little bit too complex to support for the Jinja2 template in its current shape - we can defer that to future PR.

File: agenthub/self_discover_agent/prompt.py
Comment (xingyaoww) [reply agenthub/self_discover_agent/prompt.py:77]: same here

**Review by tobitege — DISMISSED** [review]
This PR still needs to register the agent in the agenthub registry (`agenthub/__init__.py`)

File: agenthub/self_discover_agent/__init__.py
Comment (yufansong) [inline agenthub/self_discover_agent/__init__.py:1]: BTW, can you add some README.md under this folder to explain more context about self-discover agent? I also think we need some evalution result on SWE bench before merge.

File: agenthub/codeact_agent/codeact_agent.py
Comment (enyst) [reply agenthub/codeact_agent/codeact_agent.py:198]: I think so, at least last I checked. The task was not accessible to the agent via an event.

I've been thinking we could return it from `get_current_user_intent()` though. It would be easier to use `goal=state.get_user_intent()` for everything, delegate or not.

The issue is, if I recall correctly, that children start with zero history.  The task is included in the DelegateAction, but the child doesn't have access to that action. We'll need Li Boxuan to confirm, though.

File: agenthub/self_discover_agent/__init__.py
Comment (enyst) [reply agenthub/self_discover_agent/__init__.py:1]: A little curious on this, what result are we looking for?

File: agenthub/codeact_agent/codeact_agent.py
Comment (li-boxuan) [reply agenthub/codeact_agent/codeact_agent.py:198]: The contract is messy and there lacks a clear API between agents.

File: agenthub/self_discover_agent/prompt.py
Comment (sb-git-cloud) [reply agenthub/self_discover_agent/prompt.py:77]: Okay, will remove the code for now and just leave the Todo comment

File: agenthub/codeact_agent/codeact_agent.py
Comment (sb-git-cloud) [reply agenthub/codeact_agent/codeact_agent.py:198]: It seems like for now the best is to leave it as is, and once the API between agents is defined we can update the agent with a follow-up PR.

File: agenthub/self_discover_agent/agent.py
Comment (sb-git-cloud) [reply agenthub/self_discover_agent/agent.py:39]: Okay, given your comment below let's defer to a future PR.

File: agenthub/self_discover_agent/__init__.py
Comment (sb-git-cloud) [reply agenthub/self_discover_agent/__init__.py:1]: Yes, sounds good. Thanks for the suggestion. I will add a readme and run some evaluations.

**Comment by sb-git-cloud** [comment]
Thank you all for your reviews! Will work on the related changes.

**Comment by mamoodi** [comment]
Hi @sb-git-cloud thanks so much for your contribution. Just wanted to check if this is something that is generally on your radar?

**Comment by sb-git-cloud** [comment]
> Hi @sb-git-cloud thanks so much for your contribution. Just wanted to check if this is something that is generally on your radar?

Thanks for checking in. Yes. Haven't had the chance to run evaluations. Hopefully I will get to it next week.

**Comment by neubig** [comment]
Hey @sb-git-cloud , if you're interested in running evaluations, things are now a bit easier with our remote runtime: https://github.com/All-Hands-AI/OpenHands/tree/main/evaluation/swe_bench#run-inference-on-remoteruntime-experimental

If you'd like to use that we'd be happy to help out.

**Pull request closed without merging**
